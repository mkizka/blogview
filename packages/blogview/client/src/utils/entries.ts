function getEntryId(path: string) {
  return path.replace("../../entry/", "").replace(".md", "");
}

export const entries = Object.entries(
  import.meta.globEager("../../entry/*.md")
).reduce((result, [path, { default: entry }]) => {
  const entryId = getEntryId(path);
  result[entryId] = entry;
  return result;
}, {} as { [entryId: string]: string });
