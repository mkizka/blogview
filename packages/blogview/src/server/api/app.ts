import path from "path";
import express from "express";
import history from "connect-history-api-fallback";

import { entryRouter } from "./routes/entry.js";
import { configRouter } from "./routes/config.js";

const dirname = path.dirname(new URL(import.meta.url).pathname);

export function createApp() {
  const app = express();
  app.use(`/api/entry`, entryRouter);
  app.use(`/api/config`, configRouter);
  app.use(history());
  app.use(
    express.static(path.join(dirname, "..", "..", "client"), {
      setHeaders: (res) => {
        res.setHeader("Cache-Control", "no-store, no-cache, must-revalidate");
      },
    })
  );
  return app;
}
