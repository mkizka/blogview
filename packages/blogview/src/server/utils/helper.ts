import fs from "fs";
import path from "path";
import { fileURLToPath } from "node:url";

export function loadJson(filepath: string) {
  try {
    const text = fs.readFileSync(filepath, "utf-8");
    return JSON.parse(text) as Record<string, any>;
  } catch (e) {
    return null;
  }
}

const dirname = path.dirname(fileURLToPath(import.meta.url));

const pkgRootDir = path.join(dirname, "..", "..", "..");

export function pkgPath(filepath: string) {
  return path.join(pkgRootDir, filepath);
}
