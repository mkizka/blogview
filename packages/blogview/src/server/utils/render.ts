import MarkdownIt from "markdown-it";
import { HatenaPluginOptions, markdownItHatena } from "markdown-it-hatena";
import markdownItFrontMatter from "markdown-it-front-matter";
import yaml from "js-yaml";

import { BlogMeta } from "../../common/types";

type RenderMarkdownResult = {
  html: string;
  meta: BlogMeta;
};

export function renderMarkdown(
  text: string,
  options: HatenaPluginOptions
): RenderMarkdownResult {
  let meta: BlogMeta = {};
  const html = new MarkdownIt({ linkify: true, html: true })
    .use(markdownItHatena, options)
    .use(markdownItFrontMatter, (metaRaw) => {
      meta = yaml.load(metaRaw) as BlogMeta;
    })
    .render(text);
  return { html, meta };
}
