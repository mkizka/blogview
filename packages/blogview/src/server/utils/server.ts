import type { Express } from "express";
import { createServer } from "http";
import type { Server as HttpServer } from "http";
import { WebSocketServer } from "ws";
import chokidar from "chokidar";

export async function startServer(
  app: Express,
  port: number
): Promise<HttpServer> {
  const server = createServer(app);
  return new Promise((resolve, reject) => {
    server
      .listen(port)
      .once("listening", function () {
        console.log(`プレビュー: http://localhost:${port}`);
        resolve(server);
      })
      .once("error", async function (err) {
        if (err.message.includes("EADDRINUSE")) {
          console.log(`ポート${port}は既に使用されています`);
        }
        reject(err);
      });
  });
}

export async function startLocalChangesWatcher(
  server: HttpServer,
  watchPathGlob: string
) {
  const wss = new WebSocketServer({ server });
  const watcher = chokidar.watch(watchPathGlob);
  watcher.on("change", (path) => {
    console.log(`ファイルが変更されました: ${path}`);
    wss.clients.forEach((client) => client.send("Should refresh"));
  });
  process.on("SIGINT", () => {
    wss.close();
    watcher.close();
    process.exit();
  });
}
