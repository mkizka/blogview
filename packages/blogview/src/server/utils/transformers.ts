import { unified } from "unified";
import rehypeStringify from "rehype-stringify";
import remarkFrontmatter from "remark-frontmatter";
import remarkParse from "remark-parse";
import remarkStringify from "remark-stringify";
import remarkRehype from "remark-rehype";
import remarkExtract from "remark-extract-frontmatter";
import { rehypeHatenaCard } from "rehype-hatena-card";
import yaml from "yaml";

export async function md2html(md: string) {
  const processor = unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(rehypeHatenaCard)
    .use(rehypeStringify, { allowDangerousHtml: true });
  return (await processor.process(md)).value;
}

export type Frontmatter = {
  categories: string[];
  date: string;
  draft: boolean;
  id: string;
  title: string;
};

export function md2frontmatter(md: string) {
  const processor = unified()
    .use(remarkParse)
    .use(remarkStringify)
    .use(remarkFrontmatter)
    .use(remarkExtract, { yaml: yaml.parse });
  return processor.processSync(md).data as Frontmatter;
}
