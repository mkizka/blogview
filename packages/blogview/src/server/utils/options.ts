import path from "path";
import arg from "arg";

import { loadJson } from "./helper.js";

export const optionsDefault = {
  entry: "entry",
  styles: [] as string[],
  twitter: true,
  youtube: true,
};

function removeUndefinedField(obj: unknown) {
  return JSON.parse(JSON.stringify(obj)) as Record<string, any>;
}

export type Options = typeof optionsDefault;

export function getOptions(argv?: string[]) {
  const args = arg(
    {
      "--config": String,
      "-c": "--config",
      "--entry": String,
      "-e": "--entry",
      "--style": [String],
      "--no-twitter": Boolean,
      "--no-youtube": Boolean,
    },
    { permissive: true, argv }
  );

  const optionsFromArgs = removeUndefinedField({
    entry: args["--entry"],
    styles: args["--style"],
    twitter: args["--no-twitter"] && false,
    youtube: args["--no-youtube"] && false,
  });

  const optionsFromConfig = loadJson(
    path.join(process.cwd(), args["--config"] || "blogview.json")
  );

  return {
    ...optionsDefault,
    ...optionsFromConfig,
    ...optionsFromArgs,
  };
}
