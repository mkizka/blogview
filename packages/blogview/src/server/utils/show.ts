import path from "path";

import { dirname, loadJson } from "./helper.js";

const helpText = `
Usage: blogview [options]

Options:
  -h, --help    ヘルプを表示
  -v, --version バージョン名を表示
  -c, --config  設定ファイルのパスを指定(デフォルト: blogview.json)
  -e, --entry   Markdownファイルのディレクトリ指定(デフォルト: entry)
  --style       CSSファイルのパスを複数指定(デフォルト: [])
  --no-twitter  ツイート埋め込みを変換しない
  --no-youtube  YouTube埋め込みを変換しない

Details: https://github.com/mkizka/blogview#readme
`.trim();

export function showHelp() {
  console.log(helpText);
}

export function showVersion() {
  const pkg = loadJson(path.join(dirname, "..", "..", "..", "package.json"))!;
  console.log(`v${pkg.version}`);
}
