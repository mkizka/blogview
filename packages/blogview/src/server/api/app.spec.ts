import fs from "fs";
import request from "supertest";
import { describe, expect, test } from "vitest";

import { createApp } from "./app";
import { optionsDefault } from "../utils/options";
import { pkgPath } from "../utils/helper";

describe("GET /", async () => {
  test("ルートはdist/client/index.htmlを返す", async () => {
    const indexHtml = fs.readFileSync(
      pkgPath("dist/client/index.html"),
      "utf-8"
    );
    const app = createApp(optionsDefault);
    const root = await request(app).get("/");
    expect(root.text).toBe(indexHtml);
  });
});

describe("GET /api/config", () => {
  test("設定に基づいた情報を返す", async () => {
    const app = createApp({
      ...optionsDefault,
      styles: ["styles.css"],
    });
    const css = fs.readFileSync(pkgPath("styles.css"), "utf-8");
    const config = await request(app).get("/api/config");
    expect(config.body).toEqual({ styles: [css] });
  });
});

describe("GET /api/entry", () => {
  test("エントリ一覧を返す", async () => {
    const app = createApp(optionsDefault);
    const entryRoot = await request(app).get("/api/entry");
    expect(entryRoot.body.length).toBe(3);
    expect(entryRoot.body[0]).toMatchObject({
      slug: "entry1", //
      meta: {},
    });
    expect(entryRoot.body[1]).toMatchObject({
      slug: "nested/dir/entry3",
      meta: {},
    });
    expect(entryRoot.body[2]).toMatchObject({
      slug: "dir/entry2", //
      meta: {},
    });
  });
});

describe("GET /api/entry/*", async () => {
  test("個別の記事を返す", async () => {
    const app = createApp(optionsDefault);
    const entry = await request(app).get("/api/entry/entry1");
    expect(entry.body.html).toMatchSnapshot();
    expect(entry.body.meta).toMatchObject({});
  });
  test("ディレクトリ下の記事でも返す", async () => {
    const app = createApp(optionsDefault);
    const entry2 = await request(app).get("/api/entry/dir/entry2");
    const entry3 = await request(app).get("/api/entry/nested/dir/entry3");
    const noEntry = await request(app).get("/api/entry/no/entry");
    expect(entry2.ok).toBe(true);
    expect(entry3.ok).toBe(true);
    expect(noEntry.ok).toBe(false);
  });
});
