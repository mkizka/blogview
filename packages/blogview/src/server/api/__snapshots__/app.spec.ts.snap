// Vitest Snapshot v1

exports[`GET /api/entry/* > 個別の記事を返す 1`] = `
"<header class=\\"entry-header\\">
  <div class=\\"date entry-date first\\">
    <a href=\\"\\" rel=\\"nofollow\\">
      <time datetime=\\"Fri Dec 31 2021 00:32:03 GMT+0900 (Japan Standard Time)\\" title=\\"Fri Dec 31 2021 00:32:03 GMT+0900 (Japan Standard Time)\\">
        <span class=\\"date-year\\">2021</span>
        <span class=\\"hyphen\\">-</span>
        <span class=\\"date-month\\">12</span>
        <span class=\\"hyphen\\">-</span>
        <span class=\\"date-day\\">31</span>
      </time>
    </a>
  </div>
  <h1 class=\\"entry-title\\">
    <a href=\\"\\" class=\\"entry-title-link bookmark\\">エントリ1</a>
  </h1>
  <div class=\\"entry-categories categories\\">
    <a href=\\"\\" class=\\"entry-category-link category-カテゴリ1\\">カテゴリ1</a>
  </div>
</header>
<p><div class=\\"table-of-contents\\"><ul><li><ul><li><a href=\\"#%E6%AE%B5%E8%90%BD\\">段落</a></li><li><a href=\\"#%E7%94%BB%E5%83%8F%EF%BC%88%E3%81%AF%E3%81%A6%E3%81%AA%E3%83%95%E3%82%A9%E3%83%88%E3%83%A9%E3%82%A4%E3%83%95%EF%BC%89\\">画像（はてなフォトライフ）</a></li><li><a href=\\"#%E8%84%9A%E6%B3%A8\\">脚注</a></li><li><a href=\\"#%E7%B6%9A%E3%81%8D%E3%82%92%E8%AA%AD%E3%82%80\\">続きを読む</a></li><li><a href=\\"#%E7%BD%AB%E7%B7%9A\\">罫線</a></li></ul></li><li><a href=\\"#h1-%E8%A6%8B%E5%87%BA%E3%81%97\\">h1 見出し</a><ul><li><a href=\\"#h2-%E8%A6%8B%E5%87%BA%E3%81%97\\">h2 見出し</a><ul><li><a href=\\"#h3-%E8%A6%8B%E5%87%BA%E3%81%97\\">h3 見出し</a><ul><li><a href=\\"#h4-%E8%A6%8B%E5%87%BA%E3%81%97\\">h4 見出し</a><ul><li><a href=\\"#h5-%E8%A6%8B%E5%87%BA%E3%81%97\\">h5 見出し</a><ul><li><a href=\\"#h6-%E8%A6%8B%E5%87%BA%E3%81%97\\">h6 見出し</a></li></ul></li></ul></li></ul></li></ul></li><li><a href=\\"#asin\\">asin</a></li><li><a href=\\"#%E5%BC%95%E7%94%A8\\">引用</a></li><li><a href=\\"#%E3%83%AA%E3%82%B9%E3%83%88%EF%BC%88ul%2Col%2Cdl%EF%BC%89\\">リスト（ul,ol,dl）</a></li><li><a href=\\"#%E3%83%86%E3%83%BC%E3%83%96%E3%83%AB\\">テーブル</a></li><li><a href=\\"#%E3%82%B3%E3%83%BC%E3%83%89%E3%83%96%E3%83%AD%E3%83%83%E3%82%AF\\">コードブロック</a></li><li><a href=\\"#%E3%83%84%E3%82%A4%E3%83%BC%E3%83%88%E5%9F%8B%E3%82%81%E8%BE%BC%E3%81%BF\\">ツイート埋め込み</a></li><li><a href=\\"#web-%E3%82%B5%E3%82%A4%E3%83%88%E5%9F%8B%E3%82%81%E8%BE%BC%E3%81%BF\\">Web サイト埋め込み</a></li><li><a href=\\"#youtube-%E5%9F%8B%E3%82%81%E8%BE%BC%E3%81%BF\\">YouTube 埋め込み</a></li></ul></li></ul></div></p>
<h2 id=\\"%E6%AE%B5%E8%90%BD\\" tabindex=\\"-1\\">段落</h2>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<h2 id=\\"%E7%94%BB%E5%83%8F%EF%BC%88%E3%81%AF%E3%81%A6%E3%81%AA%E3%83%95%E3%82%A9%E3%83%88%E3%83%A9%E3%82%A4%E3%83%95%EF%BC%89\\" tabindex=\\"-1\\">画像（はてなフォトライフ）</h2>
<p><img src=\\"https://cdn-ak.f.st-hatena.com/images/fotolife/h/hatenablog/20101106/20101106100658.jpg\\" alt=\\"\\"></p>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<h2 id=\\"%E8%84%9A%E6%B3%A8\\" tabindex=\\"-1\\">脚注</h2>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。((ここに脚注を書きます))</p>
<h2 id=\\"%E7%B6%9A%E3%81%8D%E3%82%92%E8%AA%AD%E3%82%80\\" tabindex=\\"-1\\">続きを読む</h2>
<p>※トップページでのみ表示されます。記事ページでは表示されません。</p>
<!-- more -->
<h2 id=\\"%E7%BD%AB%E7%B7%9A\\" tabindex=\\"-1\\">罫線</h2>
<hr>
<h1 id=\\"h1-%E8%A6%8B%E5%87%BA%E3%81%97\\" tabindex=\\"-1\\">h1 見出し</h1>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。<br>
ABCDEFGHIJKLMabcdefghijklm1234567890</p>
<h2 id=\\"h2-%E8%A6%8B%E5%87%BA%E3%81%97\\" tabindex=\\"-1\\">h2 見出し</h2>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。<br>
ABCDEFGHIJKLMabcdefghijklm1234567890</p>
<h3 id=\\"h3-%E8%A6%8B%E5%87%BA%E3%81%97\\" tabindex=\\"-1\\">h3 見出し</h3>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。<br>
ABCDEFGHIJKLMabcdefghijklm1234567890</p>
<h4 id=\\"h4-%E8%A6%8B%E5%87%BA%E3%81%97\\" tabindex=\\"-1\\">h4 見出し</h4>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。<br>
ABCDEFGHIJKLMabcdefghijklm1234567890</p>
<h5 id=\\"h5-%E8%A6%8B%E5%87%BA%E3%81%97\\" tabindex=\\"-1\\">h5 見出し</h5>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。<br>
ABCDEFGHIJKLMabcdefghijklm1234567890</p>
<h6 id=\\"h6-%E8%A6%8B%E5%87%BA%E3%81%97\\" tabindex=\\"-1\\">h6 見出し</h6>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。<br>
ABCDEFGHIJKLMabcdefghijklm1234567890</p>
<h2 id=\\"asin\\" tabindex=\\"-1\\">asin</h2>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<p>[asin:4798110523:detail]</p>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<h2 id=\\"%E5%BC%95%E7%94%A8\\" tabindex=\\"-1\\">引用</h2>
<blockquote>
<p>本日（11 月 12 日）から 11 月 27 日まで、はてなダイアリーの公開デザインコンテスト「公開デザイン祭 2007 秋」を開催します。</p>
<p><cite><a href=\\"http://d.hatena.ne.jp/hatenadiary/20071112/1194858362\\">「公開デザイン祭 2007 秋」を開始しました - はてなダイアリー日記</a></cite></p>
</blockquote>
<h2 id=\\"%E3%83%AA%E3%82%B9%E3%83%88%EF%BC%88ul%2Col%2Cdl%EF%BC%89\\" tabindex=\\"-1\\">リスト（ul,ol,dl）</h2>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<ul>
<li>大賞 1 名：Amazon ギフト券 5 万円分</li>
<li>入選 9 名：Amazon ギフト券 1 万円分
<ul>
<li>大賞 1 名：Amazon ギフト券 5 万円分</li>
<li>入選 9 名：Amazon ギフト券 1 万円分</li>
</ul>
</li>
</ul>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<ol>
<li>大賞 1 名：Amazon ギフト券 5 万円分</li>
<li>入選 9 名：Amazon ギフト券 1 万円分
<ol>
<li>大賞 1 名：Amazon ギフト券 5 万円分</li>
<li>入選 9 名：Amazon ギフト券 1 万円分</li>
</ol>
</li>
</ol>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<dl>
<dt>大賞1名</dt>
<dd>Amazonギフト券5万円分</dd>
<dt>入選9名</dt>
<dd>Amazonギフト券1万円分</dd>
</dl>
<h2 id=\\"%E3%83%86%E3%83%BC%E3%83%96%E3%83%AB\\" tabindex=\\"-1\\">テーブル</h2>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<table>
<thead>
<tr>
<th>名前</th>
<th>色</th>
<th>個数</th>
</tr>
</thead>
<tbody>
<tr>
<td>りんご</td>
<td>赤</td>
<td>1</td>
</tr>
<tr>
<td>みかん</td>
<td>だいだい</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>あのイーハトヴォのすきとおった風、夏でも底に冷たさをもつ青いそら、うつくしい森で飾られたモーリオ市、郊外のぎらぎら光る草の波。</p>
<h2 id=\\"%E3%82%B3%E3%83%BC%E3%83%89%E3%83%96%E3%83%AD%E3%83%83%E3%82%AF\\" tabindex=\\"-1\\">コードブロック</h2>
<pre><code class=\\"language-css\\">.XXX {
  width: 999px;
  height: 999px;
}

.propA,
.propB,
.propC {
  width: 999px;
  height: 999px;
}

#YYY {
  width: 9px;
  height: 999px;
}
</code></pre>
<h2 id=\\"%E3%83%84%E3%82%A4%E3%83%BC%E3%83%88%E5%9F%8B%E3%82%81%E8%BE%BC%E3%81%BF\\" tabindex=\\"-1\\">ツイート埋め込み</h2>
<p><blockquote class=\\"twitter-tweet\\"><a href=\\"https://twitter.com/mkizka/status/1304351939207913472\\"></a></blockquote></p>
<h2 id=\\"web-%E3%82%B5%E3%82%A4%E3%83%88%E5%9F%8B%E3%82%81%E8%BE%BC%E3%81%BF\\" tabindex=\\"-1\\">Web サイト埋め込み</h2>
<p><iframe src=\\"https://hatenablog-parts.com/embed?url=https%3A%2F%2Fgithub.com%2F\\" scrolling=\\"no\\" frameborder=\\"0\\" style=\\"display: block; width: 100%; height: 155px; max-width: 500px; margin: 10px 0px;\\" ></iframe><cite class=\\"hatena-citation\\"> <a href=\\"https://github.com/\\">github.com</a> </cite></p>
<h2 id=\\"youtube-%E5%9F%8B%E3%82%81%E8%BE%BC%E3%81%BF\\" tabindex=\\"-1\\">YouTube 埋め込み</h2>
<p><iframe width=\\"560\\" height=\\"315\\" src=\\"https://www.youtube.com/embed/Hw5IcV3xzwY\\" frameborder=\\"0\\" allow=\\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\\" allowfullscreen></iframe></p>
"
`;
